src/back/spv/block.rs

```diff
@@ -1643,6 +1692,122 @@ impl<'w> BlockContext<'w> {
                         &self.temp_list,
                     ));
                 }
+                crate::Statement::Atomic {
+                    pointer,
+                    fun,
+                    result,
+                } => {
+                    let id = self.gen_id();
+                    let result_type_id = self.get_expression_type_id(&self.fun_info[result].ty)?;
+
+                    self.cached[result] = id;
+                    self.writer.lookup_statement_result.insert(result, id);
+
+                    let pointer_id = match self.write_expression_pointer(pointer, &mut block)? {
+                        ExpressionPointer::Ready { pointer_id } => pointer_id,
+                        ExpressionPointer::Conditional { .. } => {
+                            return Err(Error::FeatureNotImplemented(
+                                "Atomics out-of-bounds handling",
+                            ));
+                        }
+                    };
+
+                    let class = match *self.fun_info[pointer].ty.inner_with(&self.ir_module.types) {
+                        crate::TypeInner::Pointer { base: _, class } => class,
+                        _ => unimplemented!(),
+                    };
+                    let semantics =
+                        spirv::MemorySemantics::ACQUIRE_RELEASE | class.to_spirv_semantics();
+                    let scope_constant_id = self.get_scope_constant(ATOMIC_SCOPE as u32)?;
+                    let semantics_id = self.get_index_constant(semantics.bits())?;
+
+                    let instruction = match fun {
+                        crate::AtomicFunction::Binary { op, value } => {
+                            let value_id = self.cached[value];
+                            let spirv_op = match op {
+                                crate::BinaryOperator::Add => spirv::Op::AtomicIAdd,
+                                crate::BinaryOperator::And => spirv::Op::AtomicAnd,
+                                crate::BinaryOperator::InclusiveOr => spirv::Op::AtomicOr,
+                                crate::BinaryOperator::ExclusiveOr => spirv::Op::AtomicXor,
+                                _ => unimplemented!(),
+                            };
+                            Instruction::atomic_binary(
+                                spirv_op,
+                                result_type_id,
+                                id,
+                                pointer_id,
+                                scope_constant_id,
+                                semantics_id,
+                                value_id,
+                            )
```

Comment:
These `match` arms are really repetitive. We should either introduce a local function/closure, or pull the common code out after the `match`.

Reply:
slightly less repetitive now